src/index.js

```diff
@@ -0,0 +1,167 @@
+/*
+  MIT License http://www.opensource.org/licenses/mit-license.php
+  Author Tobias Koppers @sokra
+*/
+
+import validate from '@webpack-contrib/schema-utils';
+import {
+  getOptions,
+  getRemainingRequest,
+  getCurrentRequest,
+  stringifyRequest,
+} from 'loader-utils';
+import postcss from 'postcss';
+
+import schema from './options.json';
+import plugin from './plugin';
+import SyntaxError from './SyntaxError';
+
+let runtimeFile = require.resolve('./runtime');
+
+try {
+  // run test
+  runtimeFile = require.resolve('../dist/runtime');
+} catch (e) {} // eslint-disable-line no-empty
+
+export default function loader(content, map, meta) {
+  const options = getOptions(this) || {};
+
+  validate({ name: 'CSS Loader', schema, target: options });
+
+  const cb = this.async();
+  const { url, import: importOpt, sourceMap, importLoaders } = Object.assign(
+    {},
+    { url: true, import: true, sourceMap: false, importLoaders: 0 },
+    options
+  );
+
+  let prevMap = map;
+
+  if (sourceMap && prevMap) {
+    if (typeof prevMap === 'string') {
+      prevMap = JSON.parse(prevMap);
+    }
+
+    if (prevMap.sources) {
+      prevMap.sources = prevMap.sources.map((source) =>
+        source.replace(/\\/g, '/')
+      );
+      prevMap.sourceRoot = '';
+    }
+  } else {
+    // Some loaders (example `"postcss-loader": "1.x.x"`) always generates source map, we should remove it
+    prevMap = null;
+  }
+
+  let contentOrAst = content;
+
+  // Reuse CSS AST (PostCSS AST e.g 'postcss-loader') to avoid reparsing
+  if (meta) {
+    const { ast } = meta;
+
+    if (ast && ast.type === 'postcss') {
+      contentOrAst = ast.root;
+    }
+  }
+
+  const plugins = [
+    plugin({
```

Comment:
Can we split the plugin into one for `@import` and one for `url()` ?

Also passing the `loaderContext` and `importLoaders` as options makes this plugin still be coupled to the loader, if possible we should just need to loop over the messages later and call the `msg.module(result, loaderContext, ...)` `{Function}` as it's already done [here](https://github.com/webpack-contrib/css-loader/pull/748/files#diff-1fdf421c05c1140f6d71444ea2b27638R138)

Reply:
@michael-ciniawsky it is decrease perf of plugin, you can create test case there you have one plugins and two plugin and run `1000` times, you look what one plugin more faster and requires less memory.

> Also passing the loaderContext and importLoaders as options makes this plugin still be coupled to the loader, if possible we should just need to loop over the messages later and call the msg.module(result, loaderContext, ...) {Function} as it's already done here

I will look on this, maybe it is impossible before we don't drop `importLoaders` option. 

I remind you that we are not just developing pure `postcss` plugin.

Reply:
We can leave it as one it's not important, but would be cleaner 

> I will look on this, maybe it is impossible before we don't drop importLoaders option.

This is either only possible when using ES Modules or it has to do with the webpack version and was changed internally (I think between v2.0.0 => v3.0.0) , but needs to verified

> I remind you that we are not just developing pure postcss plugin.


Yes, but could be developed as if it is it's own plugin e.g `postcss-modules-import` (in theory) and the loader part is handled via messages. It's not a must and if it gets complex to do so then we simply can't do it

Reply:
@michael-ciniawsky I prefer to see good perf instead of ideal code here, we already have problem with perf here. Maybe we should drop `postcss` in favor own `tokenizer/parse` (parse only `url()/@import`) here to improve speed, but we can/should do this in next PR/version.

Reply:
I don't think this wolld have any notable perf implications, and parsing the `@import/url()` with another solution may be resulting in double parsing in the end (likely slower in total) which we are trying to void with passing the PostCSS AST now

Reply:
@michael-ciniawsky maybe, let's think about this in next major. It was just an idea.